### ISSUE TITLE
Programatically understanding Adagrad

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4079]
Programatically understanding Adagrad<br>
<br>
http://bit.ly/2t3gAGQ<br>
<br>
About 0 Minutes<br>
<br>
<div>
<div>
<div>
<div>
<h3>Or Adagrad for linear regression from basics<a href="#Or-Adagrad-for-linear-regression-from-basics">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<div>
<p>In this post, I'll be using <a href="http://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf">Adagrad</a> for solving linear regression. As usual, the purpose of this post is educational. <a href="http://ruder.io/optimizing-gradient-descent/">This link</a> gives a good overview of Adagrad alongwith other variants of Gradient Descent. To summarise from the link:</p>
<blockquote>
<p>It adapts the learning rate to the parameters, performing larger updates for infrequent and smaller updates for frequent parameters. For this reason, it is well-suited for dealing with sparse data.</p>
</blockquote>
<p>As I'd done previously, I'll be using <a href="https://github.com/HIPS/autograd">Autograd</a> to compute the gradients. Please note Autograd and not Adagrad!</p>
<h3>Formulation (<a href="https://nipunbatra.github.io/blog/2017/(http://ruder.io/optimizing-gradient-descent/">borrowed from here</a>))<a href="#Formulation-(borrowed-from-here))">&para;</a>
</h3>
<p>In regular gradient descent, we would update the <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;msup&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mi&gt;h&lt;/mi&gt;&lt;/mrow&gt;&lt;/msup&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span><span><span><span>i</span></span><span><span><span><span>t</span><span>h</span></span></span></span></span></span></span></span></span></span></span></span> parameter in the <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;+&lt;/mo&gt;&lt;msup&gt;&lt;mn&gt;1&lt;/mn&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mi&gt;h&lt;/mi&gt;&lt;/mrow&gt;&lt;/msup&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>t</span><span>+</span><span><span><span><span>1</span></span><span><span><span><span>t</span><span>h</span></span></span></span></span></span></span></span></span></span></span></span> iteration, given the learning rate <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;&amp;#x03B7;&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>&eta;</span></span></span></span></span></span></span>, where <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;msub&gt;&lt;mi&gt;g&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span><span><span><span>g</span></span><span><span><span><span>t</span><span>,</span><span>i</span></span></span></span></span></span></span></span></span></span></span></span> represents the gradient of the cost wrt <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;msup&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mi&gt;h&lt;/mi&gt;&lt;/mrow&gt;&lt;/msup&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span><span><span><span>i</span></span><span><span><span><span>t</span><span>h</span></span></span></span></span></span></span></span></span></span></span></span> param at time <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>t</span></span></span></span></span></span></span>.</p>
<p><span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML" display="block"&gt;&lt;mtable displaystyle="true"&gt;&lt;mlabeledtr&gt;&lt;mtd id="mjx-eqn-Eq 1"&gt;&lt;mtext&gt;(Eq 1)&lt;/mtext&gt;&lt;/mtd&gt;&lt;mtd&gt;&lt;msub&gt;&lt;mi&gt;&amp;#x03B8;&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;+&lt;/mo&gt;&lt;mn&gt;1&lt;/mn&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;mo&gt;=&lt;/mo&gt;&lt;msub&gt;&lt;mi&gt;&amp;#x03B8;&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;mo&gt;&amp;#x2212;&lt;/mo&gt;&lt;mi&gt;&amp;#x03B7;&lt;/mi&gt;&lt;mo&gt;&amp;#x22C5;&lt;/mo&gt;&lt;msub&gt;&lt;mi&gt;g&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;/mtd&gt;&lt;/mlabeledtr&gt;&lt;/mtable&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span><span><span><span><span><span><span><span><span><span><span><span>&theta;</span></span><span><span><span><span>t</span><span>+</span><span>1</span><span>,</span><span>i</span></span></span></span></span></span><span>=</span><span><span><span><span>&theta;</span></span><span><span><span><span>t</span><span>,</span><span>i</span></span></span></span></span></span><span>&minus;</span><span>&eta;</span><span>&sdot;</span><span><span><span><span>g</span></span><span><span><span><span>t</span><span>,</span><span>i</span></span></span></span></span></span></span></span></span></span></span></span><span><span><span><span><span>(Eq 1)</span></span></span></span></span></span></span></span></span></span></span></span></span></p>
<p>In Adagrad, we update as follows:</p>
<p><span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML" display="block"&gt;&lt;mtable displaystyle="true"&gt;&lt;mlabeledtr&gt;&lt;mtd id="mjx-eqn-Eq 2"&gt;&lt;mtext&gt;(Eq 2)&lt;/mtext&gt;&lt;/mtd&gt;&lt;mtd&gt;&lt;msub&gt;&lt;mi&gt;&amp;#x03B8;&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;+&lt;/mo&gt;&lt;mn&gt;1&lt;/mn&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;mo&gt;=&lt;/mo&gt;&lt;msub&gt;&lt;mi&gt;&amp;#x03B8;&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;mo&gt;&amp;#x2212;&lt;/mo&gt;&lt;mstyle displaystyle="true" scriptlevel="0"&gt;&lt;mfrac&gt;&lt;mi&gt;&amp;#x03B7;&lt;/mi&gt;&lt;msqrt&gt;&lt;msub&gt;&lt;mi&gt;G&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;mo&gt;+&lt;/mo&gt;&lt;mi&gt;&amp;#x03F5;&lt;/mi&gt;&lt;/msqrt&gt;&lt;/mfrac&gt;&lt;/mstyle&gt;&lt;mo&gt;&amp;#x22C5;&lt;/mo&gt;&lt;msub&gt;&lt;mi&gt;g&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;/mtd&gt;&lt;/mlabeledtr&gt;&lt;/mtable&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span><span><span><span><span><span><span><span><span><span><span><span>&theta;</span></span><span><span><span><span>t</span><span>+</span><span>1</span><span>,</span><span>i</span></span></span></span></span></span><span>=</span><span><span><span><span>&theta;</span></span><span><span><span><span>t</span><span>,</span><span>i</span></span></span></span></span></span><span>&minus;</span><span><span><span><span><span><span>&eta;</span></span><span><span><span><span><span><span><span><span><span>G</span></span><span><span><span><span>t</span><span>,</span><span>i</span><span>i</span></span></span></span></span></span><span>+</span><span>&#1013;</span></span></span><span><span><span>&oline;</span><span>&oline;</span><span>&oline;</span><span>&oline;</span><span>&oline;</span><span>&oline;</span><span>&oline;</span></span></span><span><span>&radic;</span></span></span></span></span></span></span></span></span><span>&sdot;</span><span><span><span><span>g</span></span><span><span><span><span>t</span><span>,</span><span>i</span></span></span></span></span></span></span></span></span></span></span></span><span><span><span><span><span>(Eq 2)</span></span></span></span></span></span></span></span></span></span></span></span></span></p>
<p>Here,</p>
<p><span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;msub&gt;&lt;mi&gt;G&lt;/mi&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;/mrow&gt;&lt;/msub&gt;&lt;mo&gt;&amp;#x2208;&lt;/mo&gt;&lt;msup&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi mathvariant="double-struck"&gt;R&lt;/mi&gt;&lt;/mrow&gt;&lt;mrow class="MJX-TeXAtom-ORD"&gt;&lt;mi&gt;d&lt;/mi&gt;&lt;mo&gt;&amp;#x00D7;&lt;/mo&gt;&lt;mi&gt;d&lt;/mi&gt;&lt;/mrow&gt;&lt;/msup&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span><span><span><span>G</span></span><span><span><span><span>t</span></span></span></span></span></span><span>&isin;</span><span><span><span><span><span><span>&#8477;</span></span></span></span><span><span><span><span>d</span><span>&times;</span><span>d</span></span></span></span></span></span></span></span></span></span></span></span> is a diagonal matrix where each diagonal element <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;mo&gt;,&lt;/mo&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>i</span><span>,</span><span>i</span></span></span></span></span></span></span> is the sum of the squares of the gradients w.r.t. <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;msub&gt;&lt;mi&gt;&amp;#x03B8;&lt;/mi&gt;&lt;mi&gt;i&lt;/mi&gt;&lt;/msub&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span><span><span><span>&theta;</span></span><span><span>i</span></span></span></span></span></span></span></span></span></span> up to time step <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;t&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>t</span></span></span></span></span></span></span> , while <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;&amp;#x03F5;&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>&#1013;</span></span></span></span></span></span></span> is a smoothing term that avoids division by zero (usually on the order of 1e&minus;8).</p>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Model to be learnt<a href="#Model-to-be-learnt">&para;</a>
</h3>
<p>We want to learn <code>W</code> and <code>b</code> such that:</p>
<p><span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML" display="block"&gt;&lt;mi&gt;Y&lt;/mi&gt;&lt;mo&gt;=&lt;/mo&gt;&lt;mn&gt;10&lt;/mn&gt;&lt;mi&gt;W&lt;/mi&gt;&lt;mo&gt;+&lt;/mo&gt;&lt;mi&gt;b&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>Y</span><span>=</span><span>10</span><span>W</span><span>+</span><span>b</span></span></span></span></span></span></span></p>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Defining the cost function<a href="#Defining-the-cost-function">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Dry run of cost and gradient functioning<a href="#Dry-run-of-cost-and-gradient-functioning">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<div>
<h3>Adagrad algorithm (applied on whole data batch)<a href="#Adagrad-algorithm-(applied-on-whole-data-batch)">&para;</a>
</h3>
</div>
</div>
</div>
<div>
<div>
<div>
<h4>Evolution of learning rates for <code>W</code> and <code>b</code><a href="#Evolution-of-learning-rates-for-W-and-b">&para;</a>
</h4>
</div>
</div>
</div>
<div>
<div>
<div>
<p>Ok. While There seems to be a drop in the cost, the converegence will be very slow. Remember that we had earlier found</p>
<blockquote>
<p>Cost at w=10.0, b=4.0 is: 0.0320479471939</p>
</blockquote>
<p>I'm sure this means that our parameter estimates are similar to the initial parameters and far from the true parameters. Let's just confirm the same.</p>
</div>
</div>
</div>
<div>
<div>
<div>
<p>There are a couple of interesting observations:</p>
<ol>
<li>The learning rate for <code>b</code> actually increases from its initial value of 0.01. Even after 1000 iterations, it remains more than its initial value. This can be explained by the fact that the suim of squares gradients wrt <code>b</code> would be less than 1. Thus, the denominator term by which the learning rate gets divided will be less than 1. Thus, increasing the learning rate wrt b. This can however be fixed by choosing <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;&amp;#x03F5;&lt;/mi&gt;&lt;mo&gt;=&lt;/mo&gt;&lt;mn&gt;1.0&lt;/mn&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>&#1013;</span><span>=</span><span>1.0</span></span></span></span></span></span></span>
</li>
<li>The learning rate for <code>W</code> falls very rapidly. Learning would be negligble for <code>W</code> after the initial few iterations. This can be fixed by choosing a larger initial learning rate <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;&amp;#x03B7;&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>&eta;</span></span></span></span></span></span></span>.</li>
</ol>
</div>
</div>
</div>
<div>
<div>
<div>
<h4>Evolution of <code>W</code> and <code>b</code>, wrt <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;&amp;#x03B7;&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>&eta;</span></span></span></span></span></span></span> and <span data-mathml='&lt;math xmlns="http://www.w3.org/1998/Math/MathML"&gt;&lt;mi&gt;&amp;#x03F5;&lt;/mi&gt;&lt;/math&gt;' tabindex="0"><span><span><span><span><span><span>&#1013;</span></span></span></span></span></span></span><a href="#Evolution-of-W-and-b,-wrt-%24eta%24-and-%24epsilon%24">&para;</a>
</h4>
</div>
</div>
</div>
<div>
<div>
<div>
<h5>Plotting cost v/s # Iterations<a href="#Plotting-cost-v/s-#-Iterations">&para;</a>
</h5>
</div>
</div>
</div>
<div>
<div>
<div>
<h4>Visualising the model learning<a href="#Visualising-the-model-learning">&para;</a>
</h4>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via nipunbatra.github.io http://bit.ly/2Eq3lnf<br>
<br>
May 23, 2019 at 02:38PM
